Enable WebAssembly on Python 3

turbolent opened the pull request:
- On Python 3, convert Python `bytes` to JavaScript `ArrayBuffer`s 
- Add a `to_bytes` method to JavaScript objects, which converts `TypedArray`s to Python `bytes`
- Add a test for WebAssembly

turbolent committed:
convert JS typed arrays to Python bytes, and on Python 3 bytes to JS array buffers

turbolent committed:
test Wasm on Python 3

tbodt commented:
Neat! I'll take a look tomorrow.

turbolent committed:
don't convert typed arrays by copying to Python object; need to be mutable

turbolent committed:
remove test

turbolent committed:
add a to_bytes method

tbodt commented on tests/test_convert.py:
```diff
@@ -34,9 +34,6 @@ def test_convert_to_py(context):
 
-    if sys.version_info.major >= 3:
-        assert context.eval('new Uint8Array([1, 2, 3])') == bytes([1, 2, 3])
-
```
Why remove the test?

tbodt replied:
Why remove the test?

tbodt commented on v8py/jsobject.cpp:
```diff
@@ -13,6 +13,7 @@ PyTypeObject js_object_type = {
 };
 PyMethodDef js_object_methods[] = {
     {"__dir__", (PyCFunction) js_object_dir, METH_NOARGS, NULL},
+    {"to_bytes", (PyCFunction) js_object_to_bytes, METH_NOARGS, NULL},
```
This is problematic, it leaves no way to call a javascript method on the object called `to_bytes`.

tbodt replied:
This is problematic, it leaves no way to call a javascript method on the object called `to_bytes`.

tbodt commented on v8py/convert.cpp:
```diff
@@ -141,27 +142,29 @@ Local<Value> js_from_py(PyObject *value, Local<Context> context) {
+        PyBuffer_Release(&view);
+        return hs.Escape(js_value);
+    }
+
```
What's the difference between this code and the code you deleted?

tbodt commented:
Why not use bytearray instead of bytes? That works with both python 2 and python 3, and it's mutable so the data can actually be shared.

turbolent commented:
Initially I just tried to pass `bytes` as such into V8 to instantiate a WebAssembly module (https://github.com/tbodt/v8py/pull/31/commits/ebd08f99a271e50414b23aeb91047b639e900860).
I also added the dual, to convert typed arrays back from V8 to Python `bytes` so a WebAssembly module's memory can be accessed from Python.

However, I then realized that the WebAssembly module's memory might want to be mutated, e.g. to pass data into the module, using [`TypedArray.set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/set). So instead, I removed the implicit conversion from V8 typed arrays to Python bytes in https://github.com/tbodt/v8py/pull/31/commits/de255482a7ef36cc6a765463cd4e585f645c35dd) and added the `to_bytes` method to still be able to do the conversion manually.

I guess to avoid the possible name-class for the `to_bytes` method, it could be exposed as a function from the v8py module (like `new`).

I agree that making `bytearray` work would be an even better solution, but I don't quite see how to implement that. Ideally, a slicing assignment to the bytearray in Python , e.g. `arr[1:3]`, would translate to a `TypedArray.set` call.

I looked at both the [C API for buffers](https://docs.python.org/3/c-api/buffer.html) and for [byte arrays](https://docs.python.org/dev/c-api/bytearray.html), but I'm not sure how too implement it.

I guess `__setitem__` could be implemented for `js_object`, and implemented as a `memcpy` to `ArrayBuffer.GetContents.Data`?

tbodt commented:
The V8 API supports externalized array buffers, so no need to do all that. You can just create an array buffer backed by memory owned by a python bytearray. It should also be possible to create a python bytearray or memoryview or similar backed by an array buffer.
